mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala

```diff
@@ -171,14 +177,25 @@ class GaussianMixture private (
       // Create new distributions based on the partial assignments
       // (often referred to as the "M" step in literature)
       val sumWeights = sums.weights.sum
-      var i = 0
-      while (i < k) {
-        val mu = sums.means(i) / sums.weights(i)
-        BLAS.syr(-sums.weights(i), Vectors.fromBreeze(mu),
-          Matrices.fromBreeze(sums.sigmas(i)).asInstanceOf[DenseMatrix])
-        weights(i) = sums.weights(i) / sumWeights
-        gaussians(i) = new MultivariateGaussian(mu, sums.sigmas(i) / sums.weights(i))
-        i = i + 1
+
+      if (distributeGaussians) {
+        val (ws, gs) = sc.parallelize(0 until k, math.min(k, 1024)).map { i =>
+          updateWeightsAndGaussians(sums.means(i), sums.sigmas(i), sums.weights(i), sumWeights)
```

Comment:
This might result broadcasting sums to all partitions. Could you try

~~~scala
val tuples = Seq.tabulate(i => (sums.means(i), sums.sigmas(i), sums.weights(i)))
sc.parallelize(tuples, math.min(k, 1024)).map { case (mean, sigma, weight) =>
  updateWeightsAndGaussians(mean, sigma, weight, sumWeighs)
}
~~~